File: Server/Repository/src/main/java/org/elkoserver/server/repository/RepositoryBoot.java
package org.elkoserver.server.repository;

import java.time.Clock;
import java.util.List;
import java.util.Set;
import org.elkoserver.foundation.boot.Bootable;
import org.elkoserver.foundation.net.MessageHandlerFactory;
import org.elkoserver.foundation.properties.ElkoProperties;
import org.elkoserver.foundation.server.Server;
import org.elkoserver.foundation.server.ServiceFactory;
import org.elkoserver.foundation.server.metadata.AuthDesc;
import org.elkoserver.foundation.timer.Timer;
import org.elkoserver.util.trace.Trace;
import org.elkoserver.util.trace.TraceFactory;

/**
 * The Elko boot class for the Repository.  The Repository is a server that
 * provides access to persistent storage for objects.
 */
public class RepositoryBoot implements Bootable {
    private TraceFactory traceFactory;
    private Trace tr;
    private Repository myRepository;

    public void boot(ElkoProperties props, TraceFactory traceFactory, Clock clock) {
        this.traceFactory = traceFactory;
        tr = traceFactory.trace("repo");
        Timer timer = new Timer(traceFactory, clock);
        Server server = new Server(props, "rep", tr, timer, clock, traceFactory);

        myRepository = new Repository(server, tr, traceFactory);

        if (server.startListeners("conf.listen",
                                  new RepositoryServiceFactory()) == 0) {
            tr.errori("no listeners specified");
        }
    }

    /**
     * Service factor for the Repository.
     *
     * The Repository offers two kinds of service connections:
     *
     *    repository/rep   - for object storage and retrieval
     *    repository/admin - for system administrators
     */
    private class RepositoryServiceFactory implements ServiceFactory {

        public MessageHandlerFactory provideFactory(String label,
                                                    AuthDesc auth,
                                                    Set<String> allow,
                                                    List<String> serviceNames,
                                                    String protocol)
        {
            boolean allowAdmin = false;
            boolean allowRep = false;

            if (allow.contains("any")) {
                allowAdmin = true;
                allowRep = true;
            }
            if (allow.contains("admin")) {
                allowAdmin = true;
            }
            if (allow.contains("rep")) {
                allowRep = true;
            }

            if (allowAdmin) {
                serviceNames.add("repository-admin");
            }
            if (allowRep) {
                serviceNames.add("repository-rep");
            }
            
            return new RepositoryActorFactory(myRepository, auth, allowAdmin,
                                              allowRep, tr, traceFactory);
        }
    }
}

File: Server/Repository/src/main/kotlin/org/elkoserver/server/repository/RepositoryBoot.kt
package org.elkoserver.server.repository

import org.elkoserver.foundation.boot.Bootable
import org.elkoserver.foundation.properties.ElkoProperties
import org.elkoserver.foundation.server.ShutdownWatcher
import org.elkoserver.util.trace.slf4j.Gorgel
import org.ooverkommelig.ConstantDefinition
import org.ooverkommelig.ObjectGraphConfiguration
import org.ooverkommelig.ObjectGraphLogger
import java.time.Clock

/**
 * The Elko boot class for the Repository.  The Repository is a server that
 * provides access to persistent storage for objects.
 */
@Suppress("unused")
class RepositoryBoot : Bootable {
    override fun boot(props: ElkoProperties, gorgel: Gorgel, clock: Clock) {
        val myGorgel = gorgel.getChild(RepositoryBoot::class)
        lateinit var repositoryServerGraph: RepositoryServerOgd.Graph
        val graphClosingShutdownWatcher = object : ShutdownWatcher {
            override fun noteShutdown() {
                repositoryServerGraph.close()
            }
        }
        repositoryServerGraph = RepositoryServerOgd(object : RepositoryServerOgd.Provided {
            override fun clock() = ConstantDefinition(clock)
            override fun baseGorgel() = ConstantDefinition(gorgel)
            override fun props() = ConstantDefinition(props)
            override fun externalShutdownWatcher() = ConstantDefinition(graphClosingShutdownWatcher)
        }, ObjectGraphConfiguration(object : ObjectGraphLogger {
            override fun errorDuringCleanUp(sourceObject: Any, operation: String, exception: Exception) {
                myGorgel.error("Error during cleanup of object graph. Object: $sourceObject, operation: $operation", exception)
            }
        })).Graph()
        repositoryServerGraph.server()
    }
}
